src/charm.py

```diff
@@ -0,0 +1,165 @@
+            self.on.get_postgres_password_action, self._on_get_postgres_password
+        )
+        self._cluster = PostgresqlCluster(self._get_postgresql_version())
+
+    def _on_install(self, event) -> None:
+        """Install prerequisites for the application."""
+        self.unit.status = MaintenanceStatus("installing PostgreSQL")
+
+        # Prevent the default cluster creation.
+        self._cluster.inhibit_default_cluster_creation()
+
+        # Install the PostgreSQL packages.
+        self._install_apt_packages(event, [f"postgresql-{self._get_postgresql_version()}"])
+
+        self.unit.status = WaitingStatus("waiting to start PostgreSQL")
+
+    def _on_leader_elected(self, _) -> None:
+        """Handle the leader-elected event."""
+        data = self._peers.data[self.app]
+        postgres_password = data.get("postgres-password", None)
+
+        if postgres_password is None:
+            self._peers.data[self.app]["postgres-password"] = self._new_password()
+
+    def _on_start(self, event) -> None:
+        password = self._get_postgres_password()
+        # If the leader was elected and it generated a superuser password for the all the units,
+        # the cluster can be bootstrapped in each unit.
+        if password is not None:
+            try:
+                self._cluster.bootstrap_cluster(password)
+            except ClusterAlreadyRunningError:
+                logging.error("there is already a running cluster")
+                self.unit.status = BlockedStatus("there is already a running cluster")
+            except ClusterCreateError as e:
+                logging.error("failed to create cluster")
+                self.unit.status = BlockedStatus(f"failed to create cluster with error {e}")
+            except (ClusterNotRunningError, ClusterStartError) as e:
+                logging.error("failed to start cluster")
+                self.unit.status = BlockedStatus(f"failed to start cluster with error {e}")
+            except subprocess.CalledProcessError as e:
+                logging.error("failed to bootstrap cluster")
+                self.unit.status = BlockedStatus(f"failed to bootstrap cluster with error {e}")
+            else:
+                # The cluster is up and running.
+                self.unit.status = ActiveStatus()
+        else:
+            logger.info("leader not elected and/or superuser password not yet generated")
+            self.unit.status = WaitingStatus("waiting superuser password generation")
+            event.defer()
+
+    def _on_get_postgres_password(self, event: ActionEvent) -> None:
+        """Returns the password for the postgres user as an action response."""
+        event.set_results({"postgres-password": self._get_postgres_password()})
+
+    def _get_distro_codename(self) -> str:
+        """Return the distro release code name, eg. 'bionic' or 'focal'."""
+        return subprocess.check_output(["lsb_release", "-cs"]).decode().strip()
```

Comment:
might be worth using `distro`

```python
import distro
distro.codename() # => 'focal'
```

Reply:
Good to know. I changed it to use the `distro` package.

Reply:
I removed this now as it's not needed anymore after the [update](https://github.com/canonical/operator-libs-linux/pull/35#issuecomment-993278821) in the apt library  and because the PostgresqlCluster class is now using the version retrieved through the apt library.